File: Solutionsin Haskell/p023.hs
{- 
 - Solution to Project Euler problem 23
 - Copyright (c) Sankha Saha (CODR). All rights reserved.
 - 
 - https://github.com/codr07/Project-Euler-Solusions
 -}


limit = 28123
main = putStrLn (show ans)
ans = sum (filter (not . isSumOfTwoAbundants) [1..limit])

isSumOfTwoAbundants n = any (\k -> isAbundant (n - k)) (takeWhile (< n) abundants)
abundants = filter isAbundant [1..]
isAbundant n = (divisorSum n) > n

divisorSum n = (sum [k + (div n k) | k <- upToSqrt, mod n k == 0]) - n - (if sqrt^2 == n then sqrt else 0)
	where upToSqrt = takeWhile (\k -> k * k <= n) [1..]
	      sqrt = length upToSqrt

File: codes/mathematica/p023.m
(* 
 * Solution to Project Euler problem 23
 * Copyright (c) Sankha Saha. All rights reserved.
 * 
 *  
 *   
 *)


lim = 28123;

isAbundant = Table[DivisorSigma[1, n] - n > n, {n, lim}];
abundants = Pick[Range[lim], isAbundant];
NotSumOfTwoAbundantsQ[n_] := Block[{i},
  For[i = 1, i < Length[abundants] && abundants[[i]] < n, i++,
    If[isAbundant[[n - abundants[[i]]]],
      Return[False]]];
  True]

Total[Select[Range[lim], NotSumOfTwoAbundantsQ]]
